Metricbeat Kubernetes module and metricsets docs proposal (#17636)
elastic/beats: metricbeat/docs/modules/kubernetes.asciidoc

- https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/[scheduler]
- https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/[proxy]

> Some of the previous components are running in each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.

Reviewer: ```suggestion
Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```

> For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.

Reviewer: nit 
```suggestion
For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes, Running Metricbeat on Kubernetes>> document.
```

[float]
=== Kubernetes endpoints and metricsets

Kubernetes module is a bit complex as its internal metricsets require access to a wide variety of endpoints.

> This section highlights and introduces some groups of metricsets with similar endpoint access needs. For more details on the metricsets see the ****configuration example link*** and ***metricsets reference link**** sections below.

Reviewer: configuration example link?
metricsets reference link?

Should those be replaced by sth else?

Author: @ChrsMark , I'd like to link there to the `Example Configuration` and `Metricsets` sections at the end of the final document. Take a look at the current https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-kubernetes.html

How should I add those links?

Author: Removing links and resolving conversation :)

==== node / system / pod / container / module

> The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.

Reviewer: ```suggestion
The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` on each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```

Depending on the version and configuration of Kubernetes nodes, `kubelet` might provide a read only http port (typically 10255), which is used in some configuration examples. But in general, and lately, this endpoint requires SSL (`https`) access (to port 10250 by default) and token based authentication.

==== state_* and event

All metricsets with the `state_` prefix require `hosts` field pointing to `kube-state-metrics
service` within the cluster. As the service provides cluster-wide metrics, there's no need to fetch them per node, hence the recommendation is to run these metricsets as part of a `Metricbeat Deployment` with one only replica.

> Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available [here](https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment). Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configuration.

Reviewer: Urls are provided like this:
```suggestion
Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment[here]. Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configuration.
```

Author: Changed in my local branch and pushed

==== apiserver

The apiserver metricset requires access to the Kubernetes API, which should be easily available in all Kubernetes environments. Depending on the Kubernetes configuration, the API access might require SSL (`https`) and token based authentication.

==== proxy

The proxy metricset requires access to the proxy endpoint in each of Kubernetes nodes, hence it's recommended to configure it as a part of a `Metricbeat DaemonSet`.

==== scheduler and controllermanager

> These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only in the master node, and they are not exposed via a Service, but there are different strategies available for its configuration:

Reviewer: ```suggestion
These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only on master nodes, and they are not exposed via a Service, but there are different strategies available for its configuration:
```

Author: Changed in my local branch and pushed

- Create `Kubernetes Services` to make `kube-controller-manager` and `kube-scheduler` available and configure the metricsets to point to these services as part of a Metricbeat `Deployment`.
- Use `Autodiscovery` functionality as part of a Metricbeat DaemonSet and include the metricsets in a conditional template applied for the specific pods.

> Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running in the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.

Reviewer: ```suggestion
Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running on the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```
